python_modules/dagster/dagster/_core/definitions/sensor_definition.py

```diff
@@ -978,20 +978,24 @@ def _wrapped_fn(context: SensorEvaluationContext):
         )
         result = fn(**context_param, **resource_args_populated)
 
-        if inspect.isgenerator(result) or isinstance(result, list):
-            for item in result:
-                yield item
-        elif isinstance(result, (SkipReason, RunRequest, SensorResult)):
-            yield result
-
-        elif result is not None:
-            raise Exception(
-                (
-                    "Error in sensor {sensor_name}: Sensor unexpectedly returned output "
-                    "{result} of type {type_}.  Should only return SkipReason or "
+        def check_returned_scalar(result):
+            if isinstance(result, (SkipReason, RunRequest, SensorResult)):
+                return result
+            elif result is not None:
+                raise Exception(
+                    f"Error in sensor {sensor_name}: Sensor unexpectedly returned output "
+                    f"{result} of type {type(result)}.  Should only return SkipReason or "
                     "RunRequest objects."
-                ).format(sensor_name=sensor_name, result=result, type_=type(result))
-            )
+                )
+
+        if inspect.isgenerator(result) or isinstance(result, list):
+            result_value = yield from result
+            # we reach this case if the function has both yield statements and a return statement
+            # that returns a value
+            if result_value is not None:
+                yield check_returned_scalar(result_value)
+        else:
+            yield check_returned_scalar(result)
```

Comment:
this is a little odd.. so we're basically transforming the generator into a different generator that yields the return value along with the other things that are yielded instead of returning it? (which would normally get raised as a StopIteration as I understand it - which a yield from can accept if it wants to, like we are doing here).

Maybe it would be slightly less weird if this SensorEvaluationFunction type was a boring regular function that returned a list instead of still cosplaying as a generator? It seems like it's only callsite calls list() on it right away anyways. Something about that seems slightly less odd than making it still a generator with subtly different semantics

Reply:
I implemented this change. Mind taking another look?